aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1549,20 +1558,33 @@ std::tuple<Tensor, Tensor> triangular_solve(const Tensor& self, const Tensor& A,
   TORCH_CHECK(self.dim() >= 2,
            "b should have at least 2 dimensions, but has ", self.dim(), " dimensions instead");
   TORCH_CHECK(A.dim() >= 2,
-           "u should have at least 2 dimensions, but has ", A.dim(), " dimensions instead");
+           "A should have at least 2 dimensions, but has ", A.dim(), " dimensions instead");
+
   Tensor self_broadcasted, A_broadcasted;
   std::tie(self_broadcasted, A_broadcasted) = _linalg_broadcast_batch_dims(self, A, "triangular_solve");
-  return at::_triangular_solve_helper(self_broadcasted, A_broadcasted, upper, transpose, unitriangular);
+
+  Tensor result = at::empty({0}, self.options());
+  Tensor clone_A = at::empty({0}, self.options());
+  Tensor infos = at::zeros({std::max<int64_t>(1, batchCount(self_broadcasted))}, self.options().dtype(kInt));
+
+  triangular_solve_out_info(result, clone_A, infos, A_broadcasted, self_broadcasted, upper, transpose, unitriangular);
+
+  if (self_broadcasted.dim() > 2) {
+    batchCheckErrors(infos, "triangular_solve");
+  } else {
+    singleCheckErrors(infos.item().toInt(), "triangular_solve");
+  }
+
+  return std::tuple<Tensor, Tensor>(result, clone_A);
 }
 
-std::tuple<Tensor&, Tensor&> triangular_solve_out(Tensor& result, Tensor& clone_A, const Tensor& self, const Tensor& A,
-                                                  bool upper, bool transpose, bool unitriangular) {
+std::tuple<Tensor&, Tensor&> triangular_solve_out(const Tensor& self, const Tensor& A, bool upper, bool transpose, bool unitriangular, Tensor& result, Tensor& clone_A) {
   checkSameDevice("triangular_solve", result, self);
   checkLinalgCompatibleDtype("triangular_solve", result, self);
   checkSameDevice("triangular_solve", clone_A, self, "clone_A");
   checkLinalgCompatibleDtype("triangular_solve", clone_A, self, "clone_A");
   Tensor result_tmp, clone_A_tmp;
-  std::tie(result_tmp, clone_A_tmp) = at::_triangular_solve_helper(self, A, upper, transpose, unitriangular);
+  std::tie(result_tmp, clone_A_tmp) = at::native::triangular_solve(self, A, upper, transpose, unitriangular);
   at::native::resize_output(result, result_tmp.sizes());
   at::native::resize_output(clone_A, clone_A_tmp.sizes());
   result.copy_(result_tmp);
```

Comment:
Future work: make this a safe copy

Reply:
The "safety" is achieved here with `checkSameDevice` and `checkLinalgCompatibleDtype` checks.